// @tonconnect/protocol/lib/cjs/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNode = exports.hexToByteArray = exports.toHexString = exports.splitToUint8Arrays = exports.concatUint8Arrays = exports.Base64 = void 0;
var base64_1 = require("./base64");
Object.defineProperty(exports, "Base64", { enumerable: true, get: function () { return base64_1.Base64; } });
var binary_1 = require("./binary");
Object.defineProperty(exports, "concatUint8Arrays", { enumerable: true, get: function () { return binary_1.concatUint8Arrays; } });
Object.defineProperty(exports, "splitToUint8Arrays", { enumerable: true, get: function () { return binary_1.splitToUint8Arrays; } });
Object.defineProperty(exports, "toHexString", { enumerable: true, get: function () { return binary_1.toHexString; } });
Object.defineProperty(exports, "hexToByteArray", { enumerable: true, get: function () { return binary_1.hexToByteArray; } });
var web_api_1 = require("./web-api");
Object.defineProperty(exports, "isNode", { enumerable: true, get: function () { return web_api_1.isNode; } });
